File: app/src/main/java/com/raju/javabaseproject/ui/fragments/UserFragment.java
package com.raju.javabaseproject.ui.fragments;

import android.content.Context;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.view.View;

import com.raju.javabaseproject.R;
import com.raju.javabaseproject.mvp.model.ServerResponse;
import com.raju.javabaseproject.mvp.presenter.UserPresenter;
import com.raju.javabaseproject.mvp.view.IUsersView;
import com.raju.javabaseproject.ui.fragments.base.BaseFragment;
import dagger.android.support.AndroidSupportInjection;

/**
 * Created by Rajashekhar Vanahalli on 09/04/18.
 */
public class UserFragment extends BaseFragment<UserPresenter> implements IUsersView {

    public UserFragment() {
        super(R.layout.fragment_user, "User");
    }

    public static UserFragment newInstance() {
        return new UserFragment();
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(false);
    }

    @Override
    public void onAttach(Context context) {
        AndroidSupportInjection.inject(this);
        super.onAttach(context);
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
    }

    @Override
    public void onStart() {
        super.onStart();
        /*if (presenter != null) {
            presenter.init(this);
            presenter.getHomeData();
        }*/
    }

    @Override
    public void onStop() {
        super.onStop();
        presenter.destroy();
    }

    @Override
    public void showLoading() {

    }

    @Override
    public void hideLoading() {

    }

    @Override
    public void showError(String message) {

    }

    @Override
    public void bindData(ServerResponse serverResponse) {

    }

    @Override
    public Context getContext() {
        return getActivity();
    }
}

File: app/src/main/java/com/raju/javabaseproject/ui/fragments/UserFragment.kt
package com.raju.javabaseproject.ui.fragments

import androidx.lifecycle.Observer
import android.content.Context
import android.os.Bundle
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import android.view.View

import com.raju.javabaseproject.R
import com.raju.javabaseproject.data.model.User
import com.raju.javabaseproject.ui.adapters.UserAdapter
import com.raju.javabaseproject.ui.adapters.base.ClickableItemTarget
import com.raju.javabaseproject.ui.adapters.base.ItemClickListener
import com.raju.javabaseproject.ui.adapters.delegate.base.DelegatingAdapter
import com.raju.javabaseproject.ui.fragments.base.BaseDaggerFragment
import com.raju.javabaseproject.viewmodels.UserViewModel

import javax.inject.Inject

import butterknife.BindView
import dagger.android.support.AndroidSupportInjection
import timber.log.Timber

class UserFragment : BaseDaggerFragment<UserViewModel>(R.layout.fragment_user, "User"), ItemClickListener<User> {

    @BindView(R.id.recycler_view)
    lateinit var recyclerView: RecyclerView

    @Inject
    lateinit var adapter: UserAdapter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(false)
    }

    override fun onAttach(context: Context) {
        AndroidSupportInjection.inject(this)
        super.onAttach(context)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        setupRecyclerView()
        setupAdapter(recyclerView!!)

        viewModel!!.result.observe(this, Observer {
            hideLoading()
            adapter!!.refactorItems(it)
            recyclerView!!.visibility = View.VISIBLE
        })

        viewModel!!.error.observe(this, Observer { s ->
            hideLoading()
            showSnackBar(s)
        })
    }

    override fun onStart() {
        super.onStart()
        if (viewModel != null) {
            viewModel!!.init()
            viewModel!!.getUsers()
        }
    }

    override fun onStop() {
        super.onStop()
        viewModel!!.destroy()
    }

    override fun onItemClick(view: View, position: Int, item: User) {
       Timber.d("onItemClick")
    }

    fun showLoading() {
        Timber.d("showLoading")
    }

    fun hideLoading() {
        Timber.d("hideLoading")
    }

    private fun setupRecyclerView() {
        val layoutManager = LinearLayoutManager(activity)
        recyclerView!!.layoutManager = layoutManager
    }

    private fun setupAdapter(recyclerView: RecyclerView) {
        if (recyclerView.adapter !== adapter) {
            recyclerView.adapter = adapter

            if (adapter is ClickableItemTarget<*>) {
                val target = adapter as ClickableItemTarget<User>?
                target!!.setOnItemClickListener(this)
            }

            if (adapter is DelegatingAdapter<*>) {
                adapter!!.setup()
            }
        }
    }

    companion object {

        fun newInstance(): UserFragment {
            return UserFragment()
        }
    }
}